docs: anon doc update (#4118)
neondatabase/website: content/docs/workflows/data-anonymization.md

---
title: Data anonymization
subtitle: Mask sensitive data in development branches using Postgres Anonymizer
redirectFrom:
  - /docs/concepts/anonymized-data
tag: new
enableTableOfContents: true
updatedOn: '2025-10-21T09:00:11.984Z'
---

<FeatureBeta />

Need to test against production data without exposing sensitive information? Anonymized branches let you create development copies with masked PII, emails, phone numbers, and other sensitive data.

> Neon uses [Postgres Anonymizer](https://postgresql-anonymizer.readthedocs.io/) for static data masking, and applies masking rules when you create or update the branch. This approach gives you realistic test data while protecting user privacy and supporting compliance requirements like GDPR.

Reviewer: ```suggestion
Neon uses [PostgreSQL Anonymizer](https://postgresql-anonymizer.readthedocs.io/) for static data masking, and applies masking rules when you create or update the branch. This approach gives you realistic test data while protecting user privacy and supporting compliance requirements like GDPR.
```

Author: Thanks for these, Dan!

**Key characteristics:**

- **Static masking**: Data is masked once during branch creation or when you rerun anonymization
- **Postgres Anonymizer integration**: Uses the [Postgres Anonymizer extension's](/docs/extensions/postgresql-anonymizer) masking functions

> - **Branch-specific rules**: Define different masking rules for each anonymized branch

Reviewer: ```suggestion
- **Branch-specific rules**: Define different masking rules for each anonymized branch
```
```suggestion
- **Branch-specific rules**: You can define different masking rules for each anonymized Neon branch
```

Author: Thanks!

<Admonition type="info" title="Static versus dynamic masking">
This feature uses **static masking**, which permanently transforms data in the branch when anonymization runs. Unlike dynamic masking (which masks data during queries), static masking creates an actual masked copy of the data. To get fresh data from the parent, create a new anonymized branch.
</Admonition>

## Create a branch with anonymized data

<Tabs labels={["Console", "API"]}>

<TabItem>

Select **Anonymized data** as the data option when creating a new branch.

1. Navigate to your project in the Neon Console
2. Select **Projects** -> **Branches** from the sidebar
3. Click **New Branch**
4. In the **Create new branch** dialog:
   - Select your **Parent branch** (typically `production` or `main`)
   - (Optional) Enter a **Branch name**

>    - (Optional) Set **Expire branch after** if you want automatic cleanup

Reviewer: ```suggestion
   - (Optional) Set **Expire branch after** if you want to delete the branch automatically after a specified number of days
```

  -H 'Authorization: Bearer $NEON_API_KEY' \
  -H 'Accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
    "masking_rules": [
      {
        "database_name": "neondb",
        "schema_name": "public",
        "table_name": "users",
        "column_name": "email",
        "masking_function": "anon.dummy_free_email()"
      }
    ],
    "start_anonymization": true
  }'
```

**Request parameters:**

- `masking_rules` (optional): Array of masking rules to apply to the branch. Each rule specifies:
  - `database_name`: Target database
  - `schema_name`: Target schema (typically `public`)
  - `table_name`: Table containing sensitive data
  - `column_name`: Column to mask
  - `masking_function`: Postgres Anonymizer function to apply
- `start_anonymization` (optional): Set to `true` to automatically start anonymization after branch creation

The API supports all Postgres Anonymizer masking functions, providing more options than the Console UI. You can also export and import masking rules to manage them outside of Neon.

</TabItem>

</Tabs>

## Manage masking rules

<Tabs labels={["Console", "API"]}>

<TabItem>

From the **Data Masking** page:

> 1. Select the Schema, Table, and Column you wish to mask.

Reviewer: ```suggestion
1. Select the schema, table, and column you want to mask.
```

![Neon Console 'data masking' dialog with example masking functions configured](/docs/workflows/anon-data-masking.png)

</TabItem>

<TabItem>

**Get masking rules**

```bash
GET /projects/{project_id}/branches/{branch_id}/masking_rules
```

Retrieves all masking rules defined for the branch.

**Update masking rules**

```bash
PATCH /projects/{project_id}/branches/{branch_id}/masking_rules
```

Updates masking rules for the branch. After updating rules, use the start anonymization endpoint to apply the changes.

**Start anonymization**

```bash
POST /projects/{project_id}/branches/{branch_id}/anonymize
```

Starts or restarts the anonymization process for branches in `initialized`, `error`, or `anonymized` state.

**Get anonymization status**

```bash
GET /projects/{project_id}/branches/{branch_id}/anonymized_status
```

Returns the current state (`created`, `initialized`, `initialization_error`, `anonymizing`, `anonymized`, or `error`) and progress information.

</TabItem>

</Tabs>

## Common workflow

1. Create an anonymized branch from your production branch.
2. Define masking rules for sensitive columns (emails, names, addresses, etc.).
3. Apply the masking rules.

> 4. Connect your development environment to the anonymized branch.

Reviewer: Mention where you can grab the connection string?

4. You can update rules and rerun anonymization on the branch as needed.

The parent branch data remains unchanged. Rerunning anonymization applies rules to the branch's current (already masked) data, not fresh data from the parent.

<Admonition type="note">
The branch is unavailable for connections while anonymization is in progress.
</Admonition>

## Limitations

- Cannot reset to parent, restore, or delete the read-write endpoint for anonymized branches.
- Rerunning anonymization works on already-masked data. Create a new branch for fresh parent data.
- Branch is unavailable during anonymization.
- Masking does not enforce database constraints (e.g., primary keys can be masked as NULL).
- The Console provides a curated subset of masking functions, use the API for all [Postgres Anonymizer masking functions](https://postgresql-anonymizer.readthedocs.io/en/latest/masking_functions/).

## Related resources

- [Postgres Anonymizer documentation](https://postgresql-anonymizer.readthedocs.io/)
- [Neon branching overview](/docs/introduction/branching)
- [Neon API reference](https://api-docs.neon.tech/reference/)

---

> ## Automate data anonymization with GitHub Actions

Reviewer: Maybe have this a separate page?

      }
    ],
    "start_anonymization": true
  }'
```

**Request parameters:**

- `masking_rules` (optional): Array of masking rules to apply to the branch. Each rule specifies:
  - `database_name`: Target database
  - `schema_name`: Target schema (typically `public`)
  - `table_name`: Table containing sensitive data
  - `column_name`: Column to mask
  - `masking_function`: PostgreSQL Anonymizer function to apply
- `start_anonymization` (optional): Set to `true` to automatically start anonymization after branch creation

The API supports all PostgreSQL Anonymizer masking functions, providing more options than the Console UI. You can also export and import masking rules to manage them outside of Neon.

</TabItem>

</Tabs>

## Manage masking rules

<Tabs labels={["Console", "API"]}>

<TabItem>

From the **Data Masking** page:

1. Select the schema, table, and column you want to mask.
2. Choose a masking function from the dropdown list (e.g., "Dummy Free Email" to execute `anon.dummy_free_email()`). The Console provides a curated list of common functions. For the full set of PostgreSQL Anonymizer functions, you must use the API.
3. Repeat for all sensitive columns.

> 4. When you are ready, click `Apply Masking Rules` to start the anonymization job. You can monitor its progress on this page or via the [API](#get-anonymization-status).

Reviewer 2: ```suggestion
4. When you are ready, click `Apply masking rules` to start the anonymization job. You can monitor its progress on this page or via the [API](#get-anonymization-status).
```

Author: Thanks!

3. Apply the masking rules.
4. [Connect](/docs/connect/connect-from-any-app) your development environment to the anonymized branch.
5. When you need fresh data, create a new anonymized branch.

## How anonymization works

When you create a branch with anonymized data:

1. Neon creates a new branch by copying schema and data from the parent branch.
2. You define masking rules for tables and columns containing sensitive data:
   - **Console**: The Data Masking page opens automatically after branch creation.
   - **API**: Include masking rules in the creation request or add them later via the masking rules endpoint.
3. You apply the masking rules (in Console, click **Apply Masking Rules**), and the PostgreSQL Anonymizer extension masks the branch data.
4. You can update rules and rerun anonymization on the branch as needed.

The parent branch data remains unchanged. Rerunning anonymization applies rules to the branch's current (already masked) data, not fresh data from the parent.

<Admonition type="note">
The branch is unavailable for connections while anonymization is in progress.
</Admonition>

## Limitations

> - Cannot reset to parent, restore, or delete the read-write endpoint for anonymized branches.

Reviewer 2: should we mark that we work on this, and it's a temporary restriction

**Key characteristics:**

- **Static masking**: Data is masked once during branch creation or when you rerun anonymization
- **PostgreSQL Anonymizer integration**: Uses the [PostgreSQL Anonymizer extension's](/docs/extensions/postgresql-anonymizer) masking functions
- **Branch-specific rules**: You can define different masking rules for each anonymized Neon branch

<Admonition type="info" title="Static versus dynamic masking">
This feature uses **static masking**, which permanently transforms data in the branch when anonymization runs. Unlike dynamic masking (which masks data during queries), static masking creates an actual masked copy of the data. To get fresh data from the parent, create a new anonymized branch.
</Admonition>

## Create a branch with anonymized data

<Tabs labels={["Console", "API"]}>

<TabItem>

Select **Anonymized data** as the data option when creating a new branch.

1. Navigate to your project in the Neon Console
2. Select **Projects** -> **Branches** from the sidebar
3. Click **New Branch**
4. In the **Create new branch** dialog:
   - Select your **Parent branch** (typically `production` or `main`)
   - (Optional) Enter a **Branch name**

>    - (Optional) Set **Expire branch after** if you want to delete the branch automatically after a specified number of days

Reviewer 3: We are going to update this text to "Automatically delete branch after" soon, so it'll be more clear what exactly happens to the branch.

Author: Understood, although that appears as part of a larger task (make all branches TTL 1 day) but once closer I'll update all instances in the docs.

   - Under data options, select **Anonymized data**
5. Click **Create**

> After creation, the Console loads the **Data Masking** page where you define and execute anonymization rules for your branch.

Author: Thanks! That text was moved.
